planner/cascades/transformation_rules.go

```diff
@@ -1342,6 +1346,118 @@ func (r *MergeAggregationProjection) OnTransform(old *memo.ExprIter) (newExprs [
 	return []*memo.GroupExpr{newAggExpr}, true, false, nil
 }
 
+// EliminateSingleMaxMin tries to convert a single max/min to Limit+Sort operators.
+type EliminateSingleMaxMin struct {
+	baseRule
+}
+
+// NewRuleEliminateSingleMaxMin creates a new Transformation EliminateSingleMaxMin.
+// The pattern of this rule is `max/min->X`.
+func NewRuleEliminateSingleMaxMin() Transformation {
+	rule := &EliminateSingleMaxMin{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandAggregation,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandAny, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+func (r *EliminateSingleMaxMin) Match(expr *memo.ExprIter) bool {
+	// Use appliedRuleSet in GroupExpr to avoid re-apply rules.
+	if expr.GetExpr().HasAppliedRule(r) {
+		return false
+	}
+
+	agg := expr.GetExpr().ExprNode.(*plannercore.LogicalAggregation)
+	// EliminateSingleMaxMin only works on the complete mode.
+	if !agg.IsCompleteModeAgg() {
+		return false
+	}
+	if len(agg.GroupByItems) != 0 {
+		return false
+	}
+
+	// Only one max() or min() in the AggFuncs slice, and
+	// the other aggregate functions in the AggFuncs slice should be FirstRow().
```

Comment:
Can we really elimate first row in this case? After transformation, there is a selection between source and agg, so the first row becomes `first row that the max/min arguments is not null`, is it expected?

Reply:
It's expected. See the below experiment for example.
```
mysql> select * from t;
+------+------+
| a    | b    |
+------+------+
| NULL |    1 |
|    2 |    1 |
+------+------+
2 rows in set (0.00 sec)

mysql> select max(a), b from t;
+--------+------+
| max(a) | b    |
+--------+------+
|      2 |    1 |
+--------+------+
1 row in set (0.01 sec)
```

Reply:
It's expected. In mysql5.7:
```
mysql> select * from t;
+------+------+
| a    | b    |
+------+------+
| NULL |    1 |
|    1 |    2 |
+------+------+
2 rows in set (0.00 sec)

mysql> select max(a), b from t;
+--------+------+
| max(a) | b    |
+--------+------+
|      1 |    1 |
+--------+------+
1 row in set (0.00 sec)
```

Reply:
It seems it's not easy to include the ```first_row()``` aggFun in this transformation rule. What's your opinion? @francis0407  @lamxTyler @zz-jason

Reply:
@lamxTyler good catch ！
I think we can revert this new feature. It’s hard to do that in the current framework. Just transform max/min like what we did before. 
cc @zz-jason

Reply:
I think we can rewrite the `Match()` condition to this:
1. only one `max` or `min`, we can apply this transformation
2. only one `max` or `min`, and all the `max` or `min` arguments are not null, we can apply this transformation as well.

BTW, how do calcite handle this problem?

Reply:
Another problem is that currently we cannot check if the arguments are not null. `mysql.HashNotNullFlag` does not work here, since operators like OuterJoin may generate null values on `not null columns`.

Reply:
We can add a TODO here, waiting for the not null prop is manitained.